### PR TITLE
HAAR-2186: Expand existing endpoints for authorization code

### PR DESCRIPTION
**Pull request by ravmoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/data/model/Client.kt [hunk 1]
```diff
@@ -56,6 +57,12 @@ data class Client(
   @OneToMany
   @JoinColumn(name = "registeredClientId")
   val latestClientAuthorization: MutableSet<Authorization>?,
+
+  var jwtFields: String?,
+
+  var mfaRememberMe: Boolean,
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/data/model/Client.kt:63]: Discussed & agreed OK

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/ClientsController.kt [hunk 2]
```diff
@@ -166,6 +167,10 @@ data class ClientViewResponse(
   val validDays: Long?,
   val accessTokenValidityMinutes: Long?,
   val deployment: ClientDeploymentDetails?,
+  val jwtFields: String?,
+  val mfaRememberMe: Boolean,
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/ClientsController.kt:171]: Discussed & agreed OK

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/ClientsController.kt [hunk 3]
```diff
@@ -196,14 +201,18 @@ data class ClientRegistrationRequest(
...
   val scopes: List<String>?,
   val authorities: List<String>?,
   val ips: List<String>?,
   val jiraNumber: String?,
   val databaseUserName: String?,
   val validDays: Long?,
   val accessTokenValidityMinutes: Long?,
+  val redirectUris: String?,
+  val jwtFields: String?,
+  val mfaRememberMe: Boolean,
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/ClientsController.kt:214]: Discussed & agreed OK

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/MigrationController.kt [hunk 4]
```diff
@@ -55,6 +56,10 @@ class MigrationClientRequest(
   val clientEndDate: LocalDate?,
   var lastAccessed: Instant?,
   val clientSecret: String,
-  val grantType: String,
+  val grantType: GrantType,
   val clientDeploymentDetails: ClientDeploymentDetails?,
+  val jwtFields: String?,
+  val mfaRememberMe: Boolean,
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/MigrationController.kt:62]: Discussed & agreed OK

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/ClientsController.kt [hunk 5]
```diff
@@ -213,7 +222,8 @@ data class ClientRegistrationResponse(
   val base64ClientSecret: String,
 )
 
-enum class GrantType {
-  CLIENT_CREDENTIALS,
-  AUTHORIZATION_CODE,
+// @Suppress("ktlint:standard:enum-entry-name-case")
```
Comment (ravmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/ClientsController.kt]: removed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/ClientsController.kt [hunk 6]
```diff
@@ -55,7 +56,7 @@ class ClientsController(
     clientDetails: ClientRegistrationRequest,
   ): ResponseEntity<Any> {
     val registrationResponse = clientsService.addClient(clientDetails)
-    val telemetryMap = mapOf("username" to authenticationFacade.currentUsername!!, "clientId" to clientDetails.clientId!!)
+    val telemetryMap = mapOf("username" to authenticationFacade.currentUsername!!, "clientId" to clientDetails.clientId!!, "grantType" to clientDetails.grantType.name)
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/ClientsController.kt:59]: Is the clientId really nullable here? Feels like it should not be if it is?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/ClientsController.kt (hunk 5 above)
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/ClientsController.kt]: Delete commented out annotation

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/converter/ClientConverter.kt [hunk 7]
```diff
@@ -23,14 +22,18 @@ class ClientConverter(
         clientSecretExpiresAt = null,
         clientName = clientIdService.toBase(clientId),
         clientAuthenticationMethods = CLIENT_SECRET_BASIC.value,
-        authorizationGrantTypes = CLIENT_CREDENTIALS.value,
+        authorizationGrantTypes = grantType.name.lowercase(),
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/converter/ClientConverter.kt:25]: Can't you use the description field in GrantType here instead?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/converter/MigrationClientConverter.kt [hunk 8]
```diff
@@ -23,7 +23,7 @@ class MigrationClientConverter(
         clientSecretExpiresAt = clientEndDate?.atStartOfDay()?.toInstant(ZoneOffset.UTC),
         clientName = clientIdService.toBase(clientId),
         clientAuthenticationMethods = CLIENT_SECRET_BASIC.value,
-        authorizationGrantTypes = grantType,
+        authorizationGrantTypes = grantType.name.lowercase(),
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/converter/MigrationClientConverter.kt:26]: Use description field instead?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/converter/ClientConverter.kt (hunk 7 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/converter/ClientConverter.kt]: Updated

**Review by AlTurner-MOJ — APPROVED** [review]
lgtm

**Pull request merged**